File: CS/Grokking/binary-search.py
def binary_search(elements, target):
    left = 0
    right = len(elements) - 1

    while left <= right:
        middle = (left + right) // 2
        guess = elements[middle]

        if guess == target:
            return middle

        if target < guess:
            right = middle - 1
        
        if target > guess:
            left = middle + 1

    return None

print(binary_search([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 8))

File: CS/Grokking/binary-search.js
function binarySearch(elements, target) {
  let left = 0;
  let right = elements.length - 1;

  while (left <= right) {
    const middle = Math.floor((left + right) / 2);
    const guess = elements[middle];

    if (guess === target) return middle;

    if (guess < target) right = middle - 1;
    if (guess > target) left = middle + 1;
  }

  return null;
}

console.log(binarySearch([5, 6, 7, 8, 9, 10], 7));
